src/agent/onefuzz/src/monitor.rs

```diff
@@ -45,62 +42,61 @@ impl DirectoryMonitor {
     }
 
     pub fn stop(&mut self) -> Result<()> {
-        self.watcher.unwatch(self.dir.clone())?;
+        self.watcher.unwatch(&self.dir)?;
         Ok(())
     }
 
-    pub async fn next_file(&mut self) -> Option<PathBuf> {
+    pub async fn next_file(&mut self) -> Result<Option<PathBuf>> {
         loop {
-            let event = self.notify_events.recv().await;
+            let event = match self.notify_events.recv().await {
+                Some(Ok(event)) => event,
+                Some(Err(err)) => {
+                    // A low-level watch error has occurred. Treat as fatal.
+                    warn!(
+                        "error watching for new files. path = {}, error = {}",
+                        self.dir.display(),
+                        err
+                    );
+
+                    // Make sure we try to stop our `Watcher` if we return early.
+                    let _ = self.stop();
+                    return Ok(None);
+                }
+                None => {
+                    // Make sure we try to stop our `Watcher` if we return early.
+                    let _ = self.stop();
+                    return Ok(None);
+                }
+            };
 
-            if event.is_none() {
-                // Make sure we stop our `Watcher` if we return early.
-                let _ = self.stop();
-            }
+            match event.kind {
+                EventKind::Create(..) => {
+                    let path = event
```

Comment:
Events in the new API always have a `paths` field, which contains a [vector of involved paths](https://docs.rs/notify/5.0.0-pre.14/notify/event/struct.Event.html#structfield.paths), which depend on the [`AccessKind`](https://docs.rs/notify/5.0.0-pre.14/notify/event/enum.AccessKind.html). The guarantees here are undocumented, so we just take care to handle anything.

Reply:
Why only return the first path here ? Could we receive more than 1 path for each events ?

Reply:
I'll verify as I read the source, but I think there are only multiple paths to support things like `src`/`dest` when `rename()`-ing files (see: https://docs.rs/notify/5.0.0-pre.14/notify/event/enum.EventKind.html#variant.Remove). But for that, something like a tuple of `Option<PathBuf>` values would seem to make more sense, so I'll definitely double-check.

Reply:
Yeah, on all the platforms I looked at, I'm really only seeing one `PathBuf` ever get pushed to the `paths` field. There definitely isn't anything that would cause `paths` to grow arbitrarily (e.g. debounced coalescing logic, if they were batching paths to several `Create` events that happened in quick succession).